### PR TITLE
实现block atten res 前向通信优化，meta caching / block caching交给pp model管理 ,通过传输Meta信息交流上下游caching信息，从而避免冗余block的传输

### PR DESCRIPTION
**Pull request by jjkingy**
### PR Category

Distributed Strategy

### PR Types

new features

### Description

，meta caching / block caching交给pp model管理 ,通过传输Meta信息交流上下游caching信息，从而避免冗余block的传输

### 是否引起精度变化

否，前向结果没有问题

### CODE REVIEW

**Comment by jjkingy** [comment]
你是谁你为什么一下看出来这么多bug



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2026年6月18日(周四) 晚上6:18
收件人: ***@***.***&gt;;
抄送: "yong ***@***.******@***.***&gt;;
主题: Re: [PaddlePaddle/Paddle] 实现block atten res 前向通信优化，meta caching / block caching交给pp model管理 ,通过传输Meta信息交流上下游caching信息，从而避免冗余block的传输 (PR #79334)




 
@PaddlePaddle-bot requested changes on this pull request.
  
🤖 Paddle-CI-Agent | pr_review | 2026-06-18 18:14:58
  
📋 Review 摘要
 
PR 概述：为 BlockAttnRes 前向通信增加 block/meta cache，通过上下游缓存计数裁剪 blocks 传输。
 变更范围：pipeline_parallel.py、P2P metadata 收发、dict/tuple 转换辅助函数。
 影响面 Tag：[Distributed Strategy] [Performance Optimization]
 
问题
    级别 文件 概述 
    🔴 Bug python/paddle/distributed/fleet/meta_parallel/pp_utils/p2p_communication.py:777 逐 microbatch 变化的 block_cache_meta 被现有 P2P meta cache 缓存，后续 microbatch 收不到当前 meta 
  🔴 Bug python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:3285 普通 1F1B steady 分支收到 forward input 后没有保存对应 block cache meta 
  🔴 Bug python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:3749 F-then-B 调度收到 forward input 后没有保存对应 block cache meta 
  🔴 Bug python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:3993 imbalanced-memory 调度把 forward input 延迟入队，但没有把对应 meta 与 input 成对排队 
   
📝 PR 规范检查
 
标题缺少官方 Tag，且“是否引起精度变化”段落应单独填写 否 或 是，不能追加说明文本。
 
标题建议（可直接复制）：
  
[Distributed Strategy] 优化 BlockAttnRes 前向 block 传输缓存
   PR 描述建议（点击展开，可直接复制） ### PR Category <!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] --&gt; Distributed Strategy ### PR Types <!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Others ] --&gt; Performance ### Description <!-- Describe what you’ve done --&gt; 实现 BlockAttnRes 前向通信优化：由 pipeline parallel 模型侧维护 meta/block cache，通过传输 block cache meta 同步上下游缓存计数，并按差异裁剪需要发送的 blocks，减少冗余 block 传输。变更范围包括 pipeline 调度、P2P meta 通信和 dict/tuple 转换辅助函数。 验证方式：N/A（PR 描述未提供具体测试命令或结果） ### 是否引起精度变化 <!-- one of the following [ 是 | 否 ]--&gt; 否
  
总体评价
 
优化方向和影响面清晰，但当前把逐 microbatch 控制信息复用到原有 shape meta cache 中，并且多个调度路径没有把收到的 meta 与 input tensor 绑定保存，会导致下游使用旧计数裁剪/合并 blocks。建议先修复 meta 生命周期和各 scheduler 的一致性后再合入。

 
In python/paddle/distributed/fleet/meta_parallel/pp_utils/p2p_communication.py:
 &gt;          if not self._dynamic_shape:              if not self._send_recv_meta.has_send_meta:                  self._send_recv_meta.set_send_message(output_tensor)                  self._send_recv_meta.send_meta(                      output_tensor,                      _hcg.get_pipe_parallel_group(),                      reverse=reverse, +                    block_cache_meta=block_cache_meta,  
🔴 Bug block_cache_meta 会被 p2p_cache_shape 缓存掉，后续 microbatch 收不到本次 meta。
 
_send_meta 只有在 has_send_meta 为 false 时才调用 send_meta；一旦这里把 has_send_meta 置为 self._use_cache，静态 shape/cache 开启时后续调用只做 check_send_message，不会再发送新的 block_cache_meta。接收端 _recv_meta 也同样只收一次，因此 _forward_step_helper 会用第一个或上一次的 block 计数裁剪 blocks，容易少发/多发甚至错合并。
 
建议修复方式：
 将 block cache meta 从缓存的 shape/dtype metadata 中拆出来，作为每次 forward P2P 都发送/接收的控制消息；或者在 block_cache_meta is not None 时绕过 has_send_meta / has_recv_meta 的缓存，但仍保留 tensor shape cache。修复后要确保 get_recv_block_cache_meta() 返回的是当前 input tensor 对应的 meta。
 
 
In python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:
 &gt; @@ -3176,6 +3282,7 @@ def backward_async_comm(                      recv_next=recv_next,                      batch_p2p_comm=self._use_batch_p2p_comm,                      skip_check_meta=not self.training, +                    block_cache_meta=meta_to_send,  
🔴 Bug 普通 1F1B steady 分支收到的 block cache meta 没有保存给下一个 forward。
 
这个四向通信调用在 recv_prev=True 时会同时收到下一次 forward 要用的 input_tensor 及其 meta。startup 分支和 overlap 分支已经写 _recv_block_cache_meta_for_vpp[...]，但这里直接 append input tensor，_forward_step_helper 后续只能读到旧值或 None，导致 block cache 计数不同步。
 
建议修复方式：
 在 append 前保存当前接收 meta：
 self._recv_block_cache_meta_for_vpp[next_forward_virtual_pp_rank] = (     self._p2p_helper.get_recv_block_cache_meta() if recv_prev else None )
 
并补覆盖非 overlap 1F1B 的多 microbatch 回归测试。
 
 
In python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:
 &gt; @@ -3639,6 +3746,7 @@ def forward_backward_pipeline(                  recv_prev=recv_prev,                  batch_p2p_comm=self._use_batch_p2p_comm,                  skip_check_meta=not self.training, +                block_cache_meta=meta_to_send,  
🔴 Bug F-then-B 调度收到 input 后没有同步保存对应 meta。
 
这里把 meta_to_send 传给 send_forward_recv_forward，但返回的 input_tensor 被直接 append，_recv_block_cache_meta_for_vpp[next_virtual_pp_rank] 从未更新；该 scheduler 的初始 recv_forward 也没有保存 meta。因此非首 stage 在这个调度下会按 None 或旧 meta 合并 blocks。
 
建议修复方式：
 所有 recv_forward / send_forward_recv_forward(recv_prev=True) 的结果都要和 get_recv_block_cache_meta() 一起保存；最好抽一个 helper，同步 append input 和 meta，避免不同 scheduler 漏同步。
 
 
In python/paddle/distributed/fleet/meta_parallel/pipeline_parallel.py:
 &gt; @@ -3882,6 +3990,7 @@ def forward_backward_pipeline(                  recv_prev=recv_prev,                  batch_p2p_comm=self._use_batch_p2p_comm,                  skip_check_meta=not self.training, +                block_cache_meta=meta_to_send,  
🔴 Bug imbalanced-memory 调度会把 input tensor 延迟放入队列，但没有把对应 meta 一起排队。
 
这里 input_tensor 可能先进入 forward_send_recv_buffer_queue，之后再取出 append 到 input_tensors[next_forward_virtual_pp_rank]；当前代码既没有保存收到的 meta，也没有让 meta 跟 tensor 一起进入队列。即使补一个全局 per-vpp 赋值，也会在队列延迟或多 microbatch 时和实际 input 对不上。
 
建议修复方式：
 收到 forward input 时立即取 recv_meta = get_recv_block_cache_meta()，将 (input_tensor, recv_meta) 成对放入 forward_send_recv_buffer_queue，最终 append 到 input_tensors[...] 时同步写入相同 microbatch 的 meta。
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS and Android. Download it today!

**Comment by jjkingy** [comment]
有新的pr了 这个不用了
